src/Common/FitOnFhir.Common/Services/AuthStateService.cs

```diff
@@ -79,7 +105,74 @@ public AuthState CreateAuthState(HttpRequest httpRequest)
                 ExternalSystem = jwtSecurityToken.Issuer;
             }
 
-            return new AuthState(ExternalIdentifier, ExternalSystem);
+            var redirectUrl = HttpUtility.UrlDecode(httpRequest.Query[Constants.RedirectUrlQueryParameter]);
+            if (_authenticationConfiguration.ApprovedRedirectUrls.All(url => url != redirectUrl))
+            {
+                string errorMessage = $"{redirectUrl} was not found in the list of approved redirect URLs.";
+                _logger.LogError(errorMessage);
+                throw new RedirectUrlException(errorMessage);
+            }
+
+            RedirectUrl = redirectUrl;
+            State = HttpUtility.UrlDecode(httpRequest.Query[Constants.StateQueryParameter]);
+
+            return new AuthState(ExternalIdentifier, ExternalSystem, RedirectUrl, State);
+        }
+
+        public async Task<AuthState> RetrieveAuthState(string nonce, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var blobName = EnsureArg.IsNotNullOrWhiteSpace(nonce);
+
+                // Get a reference to the blob
+                BlobClient blobClient = _blobContainerClient.GetBlobClient(blobName);
+
+                // Get the AuthState
+                var response = await blobClient.DownloadAsync(cancellationToken);
```

Comment:
It looks like `DownloadAsync` is deprecated, and we should use `DownloadContentAsync`.